nengo/builder/optimizer.py

```diff
@@ -0,0 +1,934 @@
+
+    # Once no more operators with views can be merged, we try to merge
+    # operators without views and then try again merging views (because
+    # each operator merge might generate new views).
+
+    single_pass = OpMergePass(dg)
+
+    n_initial_ops = len(dg)
+    cum_duration = 0.
+    before, after = None, None
+    i = 0
+    only_merge_ops_with_view = True
+    while only_merge_ops_with_view or after < before:
+        if max_passes is not None and i >= max_passes:
+            break
+
+        i += 1
+
+        if logger.isEnabledFor(logging.DEBUG):
+            for tp, ops in iteritems(groupby(dg, type)):
+                logger.debug("%s: %d", tp, len(ops))
+
+        only_merge_ops_with_view = before is None or before != after
+        before = len(dg)
+
+        with Timer() as t:
+            single_pass(only_merge_ops_with_view)
+
+        after = len(dg)
+        logger.info(
+            "Pass %i [%s]: Reduced %i to %i operators in %fs.",
+            i, "views" if only_merge_ops_with_view else "non-views",
+            before, after, t.duration)
+
+        # Prevent optimizer from running too long if we get up diminishing
+        # returns.
+        # Note that we don't break if there was no reduction at all because
+        # in that case we want to toggle only_merge_ops_with_view which might
+        # still yield some significant reduction.
+        cum_duration += t.duration
+        mean_reduction_rate = float(n_initial_ops - after) / cum_duration
+        last_reduction_rate = float(before - after) / t.duration
+        threshold = 0.01
+        if 0. < last_reduction_rate < threshold * mean_reduction_rate:
+            logger.info(
+                "Operator reduction rate fell below {} mean reduction rate. "
+                "Stopping optimizer.".format(threshold))
+            break
+
+    # Update model signals
+    for sigdict in itervalues(model.sig):
+        for name in sigdict:
+            while sigdict[name] in single_pass.sig_replacements:
+                sigdict[name] = single_pass.sig_replacements[sigdict[name]]
+
+    # Reinitialize the model's operator list
+    del model.operators[:]
+    for op in dg:
+        model.add_op(op)
+
+
+class OpMergePass(object):
+    def __init__(self, dg):
+        self.dg = BidirectionalDAG(dg)
+        self.might_merge = set(dg)
+        self.sig_replacements = {}
+
+        self.sig2ops = WeakKeyDefaultDict(WeakSet)
+        self.base2views = WeakKeyDefaultDict(WeakSet)
+        for op in self.dg.forward:
+            for s in op.all_signals:
+                self.sig2ops[s].add(op)
+                self.base2views[s.base].add(s)
+
+        # These variables will be initialized and used on each pass
+        self.dependents = None
+        self.only_merge_ops_with_view = None
+
+        self.merged = set()
+        self.merged_dependents = set()
+        self.opinfo = OpInfo()
+
+    def __call__(self, only_merge_ops_with_view):
```

Comment:
👉 👉 nailed it, those were my reasons. Really, the whole class could be seen as one big long function, but doing it that way would be really hard to follow and be nested 5 or 6 indentation levels deep, hence why there are methods to organize things. But, the "user" (`optimize` in this case) should never call those functions, they could treat the class as though it is just one big long function, even though it does some _stuff_ inside the class.